### PR TITLE
[Neo Plugin New feature]  UnhandledExceptionPolicy on Plugin Unhandled Exception

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

This is a follow up work after https://github.com/neo-project/neo/pull/3309. This pr will stop a plugin if a plugin throws an unhandled exception, but we also allow user to specify plugins not to stop on exception (except store plugins.).

Every single plugin can specifiy their own policy when unhandled exception is throw out of the plugin:

- **StopNode**. Exception will be throw directly and not handled, this will cause the node to stop working. And this is current logic.
- **StopPlugin**. Exception will be handled by the core and the plugin will be stoped from running.
- **Ignore**. Exception will be recorded, but node and plugin will keep running.

Fixes # follow https://github.com/neo-project/neo/pull/3309

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] TestOnPluginStopOnException
- [x] TestOnPluginStopped
- [x] TestOnNodeStopOnPluginException

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by cschuchardt88 — DISMISSED** [review]

File: src/Neo/Plugins/PluginSettings.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public UnhandledExceptionPolicy ExceptionPolicy
+    {
+        get
+        {
+            var policyString = section.GetValue("UnhandledExceptionPolicy", "StopNode");
+            if (Enum.TryParse(policyString, out UnhandledExceptionPolicy policy))
+            {
+
+                return policy;
```
Comment (shargon) [inline src/Neo/Plugins/PluginSettings.cs:28]:
```suggestion
                return policy;
```

File: src/Neo/Ledger/Blockchain.cs [hunk 2]
```diff
@@ -480,6 +484,65 @@ private void Persist(Block block)
...
+                    exceptions.Add(ex);
+                }
+            })).ToList();
+
+            await Task.WhenAll(tasks);
+
+            exceptions.ForEach(e => throw e);
+        }
+
+
```
Comment (shargon) [inline src/Neo/Ledger/Blockchain.cs:545]:
```suggestion

```

File: src/Neo/Plugins/PluginSettings.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace Neo.Plugins;
+
+public abstract class PluginSettings(IConfigurationSection section)
+{
+    public UnhandledExceptionPolicy ExceptionPolicy
+    {
+        get
+        {
+            var policyString = section.GetValue("UnhandledExceptionPolicy", "StopNode");
```
Comment (shargon) [inline src/Neo/Plugins/PluginSettings.cs:24]:
```suggestion
            var policyString = section.GetValue("UnhandledExceptionPolicy", nameof(UnhandledExceptionPolicy.StopNode));
```

File: src/Neo/Plugins/PluginSettings.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace Neo.Plugins;
+
+public abstract class PluginSettings(IConfigurationSection section)
+{
+    public UnhandledExceptionPolicy ExceptionPolicy
+    {
+        get
+        {
+            var policyString = section.GetValue("UnhandledExceptionPolicy", nameof(UnhandledExceptionPolicy.StopNode));
```
Comment (cschuchardt88) [inline src/Neo/Plugins/PluginSettings.cs:24]:
```suggestion
            var policyString = section.GetValue(nameof(UnhandledExceptionPolicy), nameof(UnhandledExceptionPolicy.StopNode));
```

File: src/Plugins/TokensTracker/TokensTracker.cs [hunk 5]
```diff
@@ -56,6 +59,11 @@ protected override void Configure()
             _maxResults = config.GetValue("MaxResults", 1000u);
             _network = config.GetValue("Network", 860833102u);
             _enabledTrackers = config.GetSection("EnabledTrackers").GetChildren().Select(p => p.Value).ToArray();
+            var policyString = config.GetValue("UnhandledExceptionPolicy", "StopNode");
```
Comment (cschuchardt88) [inline src/Plugins/TokensTracker/TokensTracker.cs:62]:
```suggestion
            var policyString = config.GetValue(nameof(UnhandledExceptionPolicy), nameof(UnhandledExceptionPolicy.StopNode));
```

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]

File: src/Neo/Plugins/PluginSettings.cs (hunk 3 above)
Comment (shargon) [reply src/Neo/Plugins/PluginSettings.cs]: there are more

**Review by cschuchardt88 — APPROVED** [review]

**Comment by superboyiii** [comment]
Maybe TokenTracker should be the same default settings as Applicationlogs.

**Comment by Jim8y** [comment]
> Maybe TokenTracker should be the same default settings as Applicationlogs, they're working in similar behaviours.

will update.

File: src/Plugins/TokensTracker/TokensTracker.json [hunk 6]
```diff
@@ -4,7 +4,8 @@
     "TrackHistory": true,
     "MaxResults": 1000,
     "Network": 860833102,
-    "EnabledTrackers": [ "NEP-11", "NEP-17" ]
+    "EnabledTrackers": [ "NEP-11", "NEP-17" ],
+    "UnhandledExceptionPolicy": "Ignore"
```
Comment (Jim8y) [inline src/Plugins/TokensTracker/TokensTracker.json:8]:
```suggestion
    "UnhandledExceptionPolicy": "StopPlugin"
```

**Review by cschuchardt88 — APPROVED** [review]

**Comment by Jim8y** [comment]
Following the call of the TC, this pr will be halted for 3-6 months. Thus please don't merge this pr.

**Comment by shargon** [comment]
> Following the call of the TC, this pr will be halted for 3-6 months. Thus please don't merge this pr.

This is required if we merged the previous related one

**Comment by Jim8y** [comment]
> > Following the call of the TC, this pr will be halted for 3-6 months. Thus please don't merge this pr.
> 
> This is required if we merged the previous related one. I Didn't read it nothing related against this pr
> 
> https://github.com/neo-project/neo/pull/3309#pullrequestreview-2121384853

sure. since coredev meeting is scheduled next week, i think it should still be fine to merge this one.

**Comment by Jim8y** [comment]
@shargon can you please merge this branch to master?

**Comment by vncoelho** [comment]
@Jim8y 

Try https://github.com/neo-project/neo/compare/plugin-unhandled-exception?expand=1

**Pull request merged**